### PR TITLE
Alerting: fix a bug where a user was able to access rules through the API from namespaces he wasn't part of

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
This PR adds shortcuts to rule api to not query database when user does not have access to any namespace. 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/ngalert/api/api_prometheus.go [hunk 1]
```diff
@@ -123,6 +123,11 @@ func (srv PrometheusSrv) RouteGetRuleStatuses(c *models.ReqContext) response.Res
 		}
 		groupId, namespaceUID, namespace := r[0], r[1], r[2]
 
+		_, ok := namespaceMap[namespaceUID]
+		if !ok {
+			continue
```
Comment (filewalkwithme) [inline pkg/services/ngalert/api/api_prometheus.go:128]: Shall we log an error, like we do with the ruler?
```
srv.log.Error("namespace not visible to the user", "user", c.SignedInUser.UserId, "namespace", r.NamespaceUID, "rule", r.UID)
```
https://github.com/grafana/grafana/blob/v8.2.2/pkg/services/ngalert/api/api_ruler.go#L185

**Comment by papagian** [comment]
I think that we need also to pass `NamespaceUIDs` here:
https://github.com/grafana/grafana/blob/f06f1b4ae704cb80c698fa5992ec58f6b98cb6bc/pkg/services/ngalert/api/api_prometheus.go#L106-L109

File: pkg/services/ngalert/api/api_prometheus.go [hunk 2]
```diff
@@ -76,6 +76,11 @@ func (srv PrometheusSrv) RouteGetRuleStatuses(c *models.ReqContext) response.Res
 		return ErrResp(http.StatusInternalServerError, err, "failed to get namespaces visible to the user")
 	}
 
+	if len(namespaceMap) == 0 {
+		srv.log.Debug("User has no access to any of namespaces")
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/api/api_prometheus.go:80]:
```suggestion
		srv.log.Debug("User does not have access to any of the namespaces")
```

File: pkg/services/ngalert/api/api_ruler.go [hunk 3]
```diff
@@ -151,6 +151,12 @@ func (srv RulerSrv) RouteGetRulesConfig(c *models.ReqContext) response.Response
 	if err != nil {
 		return ErrResp(http.StatusInternalServerError, err, "failed to get namespaces visible to the user")
 	}
+	result := apimodels.NamespaceConfigResponse{}
+
+	if len(namespaceMap) == 0 {
+		srv.log.Debug("User has no access to any of namespaces")
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/api/api_ruler.go:157]:
```suggestion
		srv.log.Debug("User does not have access to any of the namespaces")
```

**Review by JohnnyQQQQ — COMMENTED** [review]
LGTM overall

> I think that we need also to pass `NamespaceUIDs` here:
> 
> https://github.com/grafana/grafana/blob/f06f1b4ae704cb80c698fa5992ec58f6b98cb6bc/pkg/services/ngalert/api/api_prometheus.go#L106-L109

Not sure if we already looked into this, if not we should check it.

**Comment by papagian** [comment]
> LGTM overall
> 
> > I think that we need also to pass `NamespaceUIDs` here:
> > https://github.com/grafana/grafana/blob/f06f1b4ae704cb80c698fa5992ec58f6b98cb6bc/pkg/services/ngalert/api/api_prometheus.go#L106-L109
> 
> Not sure if we already looked into this, if not we should check it.

This is mainly for performance improvement; given that the actual issue is fixed by the current modifications.

**Review by JohnnyQQQQ — APPROVED** [review]
> > LGTM overall
> > > I think that we need also to pass `NamespaceUIDs` here:
> > > https://github.com/grafana/grafana/blob/f06f1b4ae704cb80c698fa5992ec58f6b98cb6bc/pkg/services/ngalert/api/api_prometheus.go#L106-L109
> > 
> > 
> > Not sure if we already looked into this, if not we should check it.
> 
> This is mainly for performance improvement; given that the actual issue is fixed by the current modifications.

Thanks for the clarification about this, then this looks good to me 🚀

**Review by gotjosh — APPROVED** [review]
The changes LGTM, thanks Yuiry for spotting this so fast! However, I think we can update the tests to reflect this change and avoid any regressions.

- Within `pkg/tests/api/alerting/api_prometheus_test.go` we have an assertion called `TestPrometheusRulesPermissions`, in there, the test creates 2 folders and removes access to one verifying the payload. We can remove the second namespace and verify the response is empty to assert this change. e.g:

```
	// remove permissions from _ALL_ folders
	require.NoError(t, store.UpdateDashboardACL(1, nil))

	// make sure that folder2 is not included in the response
	{
		promRulesURL := fmt.Sprintf("http://grafana:password@%s/api/prometheus/grafana/api/v1/rules", grafanaListedAddr)
		// nolint:gosec
		resp, err := http.Get(promRulesURL)
		require.NoError(t, err)
		t.Cleanup(func() {
			err := resp.Body.Close()
			require.NoError(t, err)
		})
		b, err := ioutil.ReadAll(resp.Body)
		require.NoError(t, err)
		require.Equal(t, 200, resp.StatusCode)

		require.JSONEq(t, `
{
	"status": "success",
	"data": {
		"groups": []
	}
}`, string(b))
	}
```

- The same applies for `pkg/tests/api/alerting/api_ruler_test.go` under `TestAlertRulePermissions`.

**Comment by gotjosh** [comment]
N.B: I would have made the changes myself given the urgency of this but this PR was created based on your fork.

File: pkg/services/ngalert/api/api_prometheus.go [hunk 4]
```diff
@@ -76,6 +76,11 @@ func (srv PrometheusSrv) RouteGetRuleStatuses(c *models.ReqContext) response.Res
 		return ErrResp(http.StatusInternalServerError, err, "failed to get namespaces visible to the user")
 	}
 
+	if len(namespaceMap) == 0 {
+		srv.log.Debug("User does not have access to any of the namespaces")
+		return response.JSON(http.StatusOK, ruleResponse)
```
Comment (gerobinson) [inline pkg/services/ngalert/api/api_prometheus.go:81]: Do we want to return `HTTP 200 OK` if the user does not have access to at least one namespace?

**Comment by gerobinson** [comment]
The other option would be to change `GetOrgRuleGroups` such that `NamespaceUIDs` is a required field instead of an optional field? The issue here is there could be code elsewhere that uses `GetOrgRuleGroups` to retrieve the rule groups for an organization across all namespaces?

File: pkg/services/ngalert/api/api_prometheus.go (hunk 4 above)
Comment (gotjosh) [reply pkg/services/ngalert/api/api_prometheus.go:81]: I believe so - returning something like `400` or `404` doesn't look right in my eyes as this doesn't necessarily mean an unexpected (or erroring) state?

**Comment by armandgrillet** [comment]
We have landed the fix in https://github.com/grafana/grafana/pull/41403 due to this PR being under your remote (thus we could not update the branch), please use the `origin` one with `yuri-tceretian/` as a prefix of your branches in your next PRs.

**Pull request closed without merging**